synapseclient/api/agent_services.py

```diff
@@ -0,0 +1,259 @@
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+
+    # Request matching <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/agent/UpdateAgentSessionRequest.html>
+    request = {
+        "sessionId": id,
+        "agentAccessLevel": access_level,
+    }
+    return await client.rest_put_async(
+        uri=f"/agent/session/{id}", body=json.dumps(request)
+    )
+
+
+async def send_prompt(
+    id: str,
+    prompt: str,
+    enable_trace: bool = False,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Any]:
+    """
+    Sends a prompt to an agent starting an asyncronous job.
+
+    Arguments:
+        id: The ID of the session to send the prompt to.
+        prompt: The prompt to send to the agent.
+        enable_trace: Whether to enable trace for the prompt. Defaults to False.
+        synapse_client: If not passed in and caching was not disabled by
+                `Synapse.allow_client_caching(False)` this will use the last created
+                instance from the Synapse class constructor.
+
+    Returns:
+        The response matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/asynch/AsyncJobId.html>
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+
+    # Request matching <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/agent/AgentChatRequest.html>
+    request = {
+        "concreteType": "org.sagebionetworks.repo.model.agent.AgentChatRequest",
+        "sessionId": id,
+        "chatText": prompt,
+        "enableTrace": enable_trace,
+    }
+    return await client.rest_post_async(
+        uri="/agent/chat/async/start", body=json.dumps(request)
+    )
+
+
+async def get_response(
+    prompt_id: str,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Any]:
+    """
+    Gets the response to a prompt. If the response is not ready, the endpoint will return a reponse matching
+    <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/asynch/AsynchronousJobStatus.html>.
+    In this case, the response retrieval is retried every second until the response is ready or a timeout is reached.
+
+    Arguments:
+        prompt_id: The token of the prompt to get the response for.
+        synapse_client: If not passed in and caching was not disabled by
+                `Synapse.allow_client_caching(False)` this will use the last created
+                instance from the Synapse class constructor.
+
+    Returns:
+        The response matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/agent/AgentChatResponse.html>
+
+    Raises:
+        TimeoutError: If the response is not ready after 1 minute
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+    start_time = asyncio.get_event_loop().time()
+    TIMEOUT = 60
+
+    while True:
```

Comment:
https://github.com/Sage-Bionetworks/synapsePythonClient/blob/8bd5e195c389e3a83bd1308be3c6a42f25d87b8f/synapseclient/client.py#L5029

This function exists on the client, but it not built with asyncio in mind. There is some additional error handling it does for the failed state. I could see an asyncio compliant version being written and used here.

Maybe it could go in this file? https://github.com/Sage-Bionetworks/synapsePythonClient/blob/8bd5e195c389e3a83bd1308be3c6a42f25d87b8f/synapseclient/core/async_utils.py

Reply:
`_waitForAync` does both the POST call (starting the job) and the GET call (retrieving the result). Should the functionality I have written here work that way too? I was viewing it as 1 function per API call in agent_services originally.

Reply:
They can be 2 separate functions that handle each part. In the Agent script you'll chain them together